store/sqlstore/channel_member_history_store.go

```diff
@@ -0,0 +1,113 @@
+
+package sqlstore
+
+import (
+	"net/http"
+
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/mattermost/mattermost-server/store"
+
+	l4g "github.com/alecthomas/log4go"
+)
+
+type SqlChannelMemberHistoryStore struct {
+	SqlStore
+}
+
+func NewSqlChannelMemberHistoryStore(sqlStore SqlStore) store.ChannelMemberHistoryStore {
+	s := &SqlChannelMemberHistoryStore{
+		SqlStore: sqlStore,
+	}
+
+	for _, db := range sqlStore.GetAllConns() {
+		table := db.AddTableWithName(model.ChannelMemberHistory{}, "ChannelMemberHistory").SetKeys(false, "ChannelId", "UserId", "JoinTime")
+		table.ColMap("ChannelId").SetMaxSize(26)
+		table.ColMap("UserId").SetMaxSize(26)
+		table.ColMap("JoinTime").SetNotNull(true)
+	}
+
+	return s
+}
+
+func (s SqlChannelMemberHistoryStore) LogJoinEvent(userId string, channelId string, joinTime int64) store.StoreChannel {
+	return store.Do(func(result *store.StoreResult) {
+		channelMemberHistory := &model.ChannelMemberHistory{
+			UserId:    userId,
+			ChannelId: channelId,
+			JoinTime:  joinTime,
+		}
+
+		if err := s.GetMaster().Insert(channelMemberHistory); err != nil {
+			result.Err = model.NewAppError("SqlChannelMemberHistoryStore.LogJoinEvent", "store.sql_channel_member_history.log_join_event.app_error", map[string]interface{}{"ChannelMemberHistory": channelMemberHistory}, err.Error(), http.StatusInternalServerError)
+		}
+	})
+}
+
+func (s SqlChannelMemberHistoryStore) LogLeaveEvent(userId string, channelId string, leaveTime int64) store.StoreChannel {
+	return store.Do(func(result *store.StoreResult) {
+		query := `
+			SELECT
+				cmh.*,
+				u.Email
+			FROM ChannelMemberHistory cmh
+			INNER JOIN Users u ON cmh.UserId = u.Id
+			WHERE cmh.UserId = :UserId
+			AND cmh.ChannelId = :ChannelId
+			AND cmh.LeaveTime IS NULL
+			ORDER BY cmh.JoinTime ASC
+			LIMIT 1`
+
+		params := map[string]interface{}{"UserId": userId, "ChannelId": channelId}
+		channelMemberHistory := new(model.ChannelMemberHistory)
+		if err := s.GetReplica().SelectOne(&channelMemberHistory, query, params); err != nil {
+			// if there was no join event, we can't log a leave event, but we shouldn't fail
+			l4g.Warn("Channel join event for user %v and channel %v not found", userId, channelId)
+		} else {
+			channelMemberHistory.LeaveTime = &leaveTime
+
+			if _, err := s.GetMaster().Update(channelMemberHistory); err != nil {
+				result.Err = model.NewAppError("SqlChannelMemberHistoryStore.LogLeaveEvent", "store.sql_channel_member_history.log_leave_event.update_error", map[string]interface{}{"ChannelMemberHistory": channelMemberHistory}, err.Error(), http.StatusInternalServerError)
+			}
+		}
+	})
+}
+
+func (s SqlChannelMemberHistoryStore) GetUsersInChannelDuring(startTime int64, endTime int64, channelId string) store.StoreChannel {
+	return store.Do(func(result *store.StoreResult) {
+		query := `
+			SELECT
+				cmh.*,
+				u.Email
+			FROM ChannelMemberHistory cmh
+			INNER JOIN Users u ON cmh.UserId = u.Id
```

Comment:
What's with the INNER JOIN?

Reply:
The code needs the user's email address, but the ChannelMemberHistory table only contains the user's id. We need to join on Users to get the email address. Users can change their email addresses, so it's important that we don't include the email in the ChannelMemberHistory table, as it could change and become stale.

Reply:
👍